### PR TITLE
feat: use `vue-tsc` generate vue declarations

### PR DESCRIPTION
**Pull request by aa900031**
Use [vue-tsc](https://github.com/vuejs/language-tools/tree/master/packages/vue-tsc) to generate d.ts for Vue SFC

Reference: #22

### CODE REVIEW

**Comment by aa900031** [comment]
@danielroe Thank you for fixed lint, Sorry, I didn't notice that

**Review by danielroe — APPROVED** [review]
I think ideally I would like to refactor this to be a separate 'vue-dts' loader. But I think we can probably merge and iterate later. wdyt @pi0?

**Comment by robinscholz** [comment]
Would be great to get this merged!

**Comment by danielroe** [comment]
Thoughts @pi0?

**Review by pi0 — CHANGES_REQUESTED** [review]
Hi again thanks for this PR and sorry for checking late. I like the general idea and the implementation but checking with current status and `pnpm dev` to run on fixture, it generates additional files inside `src` + empty ones in dist. We mighe need to check why tsHost is possibly not being passed.

Also would be nice if we have a fallback strategy to use ts loader if `vue-tsc` is not installed (maybe with a warn to guide to add)

**Comment by aa900031** [comment]
@pi0 Thank you for review, the `vfs` is different after my commit  48478dffbb867302e5d0b9985ca3083c7da89128, let me check the issue later, maybe need to split to single loader ?

**Comment by pi0** [comment]
Standalone loader would be a nice idea!

**Comment by aa900031** [comment]
Hi @pi0, based on my research, `vue-tsc` no longer utilizes `options.host.writeFile` after v1.7.8. Therefore, I have overridden ts.sys.writeFile instead.

**Comment by aa900031** [comment]
@pi0 Hi again, It's been a month, could you please let me know when we can expect this to be merged?

**Comment by DevilTea** [comment]
Any update for this🥺?

**Comment by pi0** [comment]
If it is important, feel free to make a new PR to rework 👍🏼 (we need to fix the correct dist output)

**Comment by aa900031** [comment]
@pi0 Using the latest vue-tsc version 1.8.22 is not working. I found that rewriting ts.sys and the ts.sys used inside vue-tsc are different. I am currently looking for a way to fix this issue.

**Comment by aa900031** [comment]
Hi @pi0, Inside vue-tsc, `require` is used instead of `import`. In order to override `ts.sys`, it is necessary to import it in the same way as vue-tsc for them to refer to the same file, so I use same way to import typescript and override is work!

Please check that again, Thanks !

File: src/utils/vue-dts.ts [hunk 1]
```diff
@@ -30,22 +30,22 @@ export async function getVueDeclarations(
     return;
   }
 
-  const ts = await import("typescript").then((r) => r.default || r);
+  // Inside vue-tsc, `require` is used instead of `import`. In order to override `ts.sys`, it is necessary to import it in the same way as vue-tsc for them to refer to the same file.
+  const ts =
+    require("typescript") as typeof import("typescript/lib/tsserverlibrary");
```
Comment (pi0) [inline src/utils/vue-dts.ts:35]: We need to use `createRequire` from `node:module` to create a new require this require is not available in strict `.mjs`

Comment (aa900031) [reply src/utils/vue-dts.ts:38]: fix in 6a29adf172aa62934242a17fb3fe86e2b1064f58

Comment (aa900031) [reply src/utils/vue-dts.ts:38]: Hi @pi0 ! I have fixed it. Please take another look. Thank you.

**Comment by aa900031** [comment]
Hi @pi0, are there any pending questions? If not, please proceed with the merge. I’m eager to use this feature!

**Comment by aa900031** [comment]
Hi everyone, I'm not sure why the review has been pause. However, since I urgently need this feature, I have independently released `@aa900031/mkdist` v1.5.0, which includes the functionality provided by this PR.

If you need this functionality like I do, below are a few ways to use:

Using alias in package.json:

```json
// package.json

{
    "devDependencies": {
        "mkdist": "npm:@aa900031/mkdist@^1.5.0"
    }
}
```

If you use `unbuild`, you can use the package manager's overrides feature. Here's an example for pnpm:

```json
// package.json

{
    "devDependencies": {
        "unbuild": "^2.0.0"
    },
    "pnpm": {
        "overrides": {
            "unbuild>mkdist": "npm:@aa900031/mkdist@^1.5.0"
        }
    }
}
```

**Review by pi0 — APPROVED** [review]

**Comment by pi0** [comment]
Thanks and sorry for dalay on merge (other than review process, I'm really busy on many project so have to perorotize).

**Comment by aa900031** [comment]
It's ok, I know you have many projects, Thank you so much :bow:

**Pull request merged**